schema/encoding/jsonschema/encoder.go

```diff
@@ -11,57 +11,37 @@ import (
 // FieldValidator types in the schema package is supported at the moment. Custom validators are also not yet handled.
 // Attempting to encode a schema containing such fields will result in a ErrNotImplemented error.
 type Encoder struct {
-	io.Writer
+	w io.Writer
 }
 
 // NewEncoder returns a new JSONSchema Encoder that writes to w.
 func NewEncoder(w io.Writer) *Encoder {
-	return &Encoder{w}
+	return &Encoder{w: w}
 }
 
 // Encode writes the JSON Schema representation of s to the stream, followed by a newline character.
 func (e *Encoder) Encode(s *schema.Schema) error {
-	ew := errWriter{w: e.Writer}
-	ew.writeString("{")
-	if ew.err == nil {
-		ew.err = encodeSchema(e.Writer, s)
+	m := make(map[string]interface{})
+	if err := addSchemaProperties(m, s); err != nil {
+		return err
 	}
-	ew.writeString("}\n")
-	return ew.err
-}
-
-// Wrap IO writer so we can consolidate error handling in a single place. Also track properties written so we know when
-// to emit a separator.
-type errWriter struct {
-	w   io.Writer // writer instance
-	err error     // track errors
-}
+	enc := json.NewEncoder(e.w)
+	return enc.Encode(m)
 
-// Write ensures compatibility with the io.Writer interface.
-func (ew errWriter) Write(p []byte) (int, error) {
-	if ew.err != nil {
-		return 0, ew.err
-	}
-	return ew.w.Write(p)
 }
 
-func (ew errWriter) writeFormat(format string, a ...interface{}) {
-	if ew.err != nil {
-		return
-	}
-	_, ew.err = fmt.Fprintf(ew.w, format, a...)
+// The Builder interface should be implemented by custom schema.FieldValidator implementations to allow JSON Schema
+// serialization.
+type Builder interface {
+	// JSONSchema should return a map containing JSON Schema Draft 4 properties that can be set based on
+	// FieldValidator data. Application specific properties can be added as well, but should not conflict with any
+	// legal JSON Schema keys.
+	JSONSchema() (map[string]interface{}, error)
 }
 
-func (ew errWriter) writeString(s string) {
-	if ew.err != nil {
-		return
-	}
-	_, ew.err = ew.w.Write([]byte(s))
-}
+// builderFunc is an adapter that allows pure functions to implement the Builder interface.
+type builderFunc func() (map[string]interface{}, error)
```

Comment:
PS! I am happy to expose it later, if we can find a real end-user use-case for it.